src/devicecode.rs

```diff
@@ -0,0 +1,264 @@
+                "access_denied" => DeviceCodeErrorResponseType::AccessDenied,
+                "expired_token" => DeviceCodeErrorResponseType::ExpiredToken,
+                _ => DeviceCodeErrorResponseType::Basic(BasicErrorResponseType::Extension(ext)),
+            },
+            basic => DeviceCodeErrorResponseType::Basic(basic),
+        }
+    }
+}
+impl AsRef<str> for DeviceCodeErrorResponseType {
+    fn as_ref(&self) -> &str {
+        match self {
+            DeviceCodeErrorResponseType::AuthorizationPending => "authorization_pending",
+            DeviceCodeErrorResponseType::SlowDown => "slow_down",
+            DeviceCodeErrorResponseType::AccessDenied => "access_denied",
+            DeviceCodeErrorResponseType::ExpiredToken => "expired_token",
+            DeviceCodeErrorResponseType::Basic(basic) => basic.as_ref(),
+        }
+    }
+}
+impl<'de> serde::Deserialize<'de> for DeviceCodeErrorResponseType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        let variant_str = String::deserialize(deserializer)?;
+        Ok(Self::from_str(&variant_str))
+    }
+}
+impl serde::ser::Serialize for DeviceCodeErrorResponseType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        serializer.serialize_str(self.as_ref())
+    }
+}
+impl ErrorResponseType for DeviceCodeErrorResponseType {}
+impl Debug for DeviceCodeErrorResponseType {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), FormatterError> {
+        Display::fmt(self, f)
+    }
+}
+
+impl Display for DeviceCodeErrorResponseType {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), FormatterError> {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+///
+/// Error response specialization for device code OAuth2 implementation.
+///
+pub type DeviceCodeErrorResponse = StandardErrorResponse<DeviceCodeErrorResponseType>;
+
+pub(crate) struct DeviceAccessResult<TR, RE, TE, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+    RE: Error + 'static,
+{
+    res: Result<TR, RequestTokenError<RE, TE>>,
+    _phantom: PhantomData<TT>,
+}
+
+impl<TR, RE, TE, TT> DeviceAccessResult<TR, RE, TE, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+    RE: Error + 'static,
+{
+    pub fn new(res: Result<TR, RequestTokenError<RE, TE>>) -> Self {
+        Self {
+            res,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn result(self) -> Result<TR, RequestTokenError<RE, TE>> {
+        self.res
+    }
+}
+
+pub(crate) enum DeviceAccessTokenPollResult<TR, RE, TE, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+    RE: Error + 'static,
+{
+    ContinueWithNewPollInterval(Duration),
+    Done(DeviceAccessResult<TR, RE, TE, TT>),
```

Comment:
nit: I think we can get rid of the `DeviceAccessResult` struct by making this:
```
Done(Result<TR, RequestTokenError<RE, TE>>), PhantomData<TT>)
```

Reply:
Done.